test/source-updater.test.js

```diff
@@ -1,232 +1,769 @@
+  this.sourceUpdater.createSourceBuffers({
+    audio: 'mp4a.40.2'
+  });
+  this.sourceUpdater.videoTimestampOffset(999);
+
+  assert.equal(
+    this.sourceUpdater.videoTimestampOffset(), 0, 'offset stays at initial value');
+});
+
+QUnit.test('setting audio timestamp offset without buffer is a noop', function(assert) {
+  // only create the video buffer
+  this.sourceUpdater.createSourceBuffers({
+    video: 'avc1.4d400d'
+  });
+  this.sourceUpdater.audioTimestampOffset(999);
+
+  assert.equal(
+    this.sourceUpdater.audioTimestampOffset(), 0, 'offset stays at initial value');
+});
+
+QUnit.test('not ready by default', function(assert) {
+  assert.notOk(this.sourceUpdater.ready(), 'source updater is not ready');
+});
+
+QUnit.test('ready with a video buffer', function(assert) {
+  this.sourceUpdater.createSourceBuffers({
+    video: 'avc1.4d400d'
+  });
+  assert.ok(this.sourceUpdater.ready(), 'source updater is ready');
+});
+
+QUnit.test('ready with an audio buffer', function(assert) {
+  this.sourceUpdater.createSourceBuffers({
+    audio: 'mp4a.40.2'
+  });
+  assert.ok(this.sourceUpdater.ready(), 'source updater is ready');
+});
+
+QUnit.test('ready with both an audio and video buffer', function(assert) {
+  this.sourceUpdater.createSourceBuffers({
+    video: 'avc1.4d400d',
+    audio: 'mp4a.40.2'
+  });
+  assert.ok(this.sourceUpdater.ready(), 'source updater is ready');
+});
+
+QUnit.test('waits for sourceopen to create source buffers', function(assert) {
+  const mockMediaSource = {
+    addEventListenerCalls: [],
+    addSourceBufferCalls: [],
+    // readyState starts as closed, source updater has to wait for it to open
+    readyState: 'closed',
+    addEventListener:
+      (name, callback) => mockMediaSource.addEventListenerCalls.push({ name, callback }),
+    addSourceBuffer: (mimeType) => {
+      mockMediaSource.addSourceBufferCalls.push(mimeType);
+      return {
+        // source updater adds event listeners immediately after creation, mock out to
+        // prevent errors
+        addEventListener() {}
+      };
+    }
+  };
+
+  this.sourceUpdater = new SourceUpdater(mockMediaSource);
+
+  assert.equal(
+    mockMediaSource.addEventListenerCalls.length, 0, 'no event listener calls');
+  assert.equal(
+    mockMediaSource.addSourceBufferCalls.length, 0, 'no add source buffer calls');
+
+  this.sourceUpdater.createSourceBuffers({
+    video: 'avc1.4d400d',
+    audio: 'mp4a.40.2'
+  });
+
+  assert.equal(
+    mockMediaSource.addEventListenerCalls.length, 1, 'one event listener');
+  assert.equal(
+    mockMediaSource.addEventListenerCalls[0].name,
+    'sourceopen',
+    'listening on sourceopen');
+  assert.equal(
+    mockMediaSource.addSourceBufferCalls.length, 0, 'no add source buffer calls');
+
+  mockMediaSource.readyState = 'open';
+  mockMediaSource.addEventListenerCalls[0].callback();
+
+  assert.equal(
+    mockMediaSource.addEventListenerCalls.length, 1, 'one event listener');
```

Comment:
do we ever clean up the `sourceopen` listener before disposing the sourceUpdater?

Reply:
Added in https://github.com/videojs/http-streaming/pull/272/commits/3641cf6ec52129245935561cb772747aa61c88b0